Makefile

```
@@ -97,57 +62,55 @@ e2e: kind-clusters ## Runs end-to-end tests against KinD clusters
 			--istio.test.onlyWorkloads=standard; \
 	)
 
-## Tooling
-
-$(shell mkdir -p $(LOCALBIN))
-
-$(GCI):
-	@GOBIN=$(LOCALBIN) go install github.com/daixiang0/gci@v0.13.5
-
-$(HELM):
-	@curl -sSL https://get.helm.sh/helm-v3.14.2-linux-amd64.tar.gz -o $(LOCALBIN)/helm.tar.gz
-	@tar -xzf $(LOCALBIN)/helm.tar.gz -C $(LOCALBIN) --strip-components=1 linux-amd64/helm
-	@rm -f $(LOCALBIN)/helm.tar.gz
-
-$(PROTOC):
-	@curl -sSL https://github.com/protocolbuffers/protobuf/releases/download/v21.12/protoc-21.12-linux-x86_64.zip -o $(LOCALBIN)/protoc.zip
-	@python3 -c "import zipfile; z=zipfile.ZipFile('$(LOCALBIN)/protoc.zip'); z.extract('bin/protoc', '$(LOCALBIN)')"
-	@mv $(LOCALBIN)/bin/protoc $(LOCALBIN)
-	@rm -rf $(LOCALBIN)/bin
-	@rm -f $(LOCALBIN)/protoc.zip
-	@chmod +x $(PROTOC)
-
-$(PROTOC_GEN_GO):
-	@GOBIN=$(LOCALBIN) go install -mod=readonly google.golang.org/protobuf/cmd/protoc-gen-go@v1.36.0
+.PHONY: kind-clusters
+kind-clusters: $(KIND) $(HELM) ## Provisions KinD clusters for local development or testing
+	@local_tag=$(call local_tag); \
+	$(MAKE) docker-build -e TAG=$$local_tag; \
+	export TAG=$$local_tag; \
+	PATH=$(LOCALBIN):$$PATH \
+	$(PROJECT_DIR)/test/scripts/kind_provisioner.sh
 
-$(PROTOC_GEN_GRPC):
-	@GOBIN=$(LOCALBIN) go install -mod=readonly google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.5.1
+##@ Containers
 
-$(PROTOC_GEN_DEEPCOPY):
-	@GOBIN=$(LOCALBIN) go install -mod=readonly istio.io/tools/cmd/protoc-gen-golang-deepcopy@latest
+CONTAINER_CLI ?= docker
+## Image settings need to be exported.
+## KinD scripts rely on them to determine which images should be used
+## and if they should be pushed to node's repository (USE_LOCAL_IMAGE).
+export USE_LOCAL_IMAGE ?= true
+export TAG ?= latest
+export HUB ?= quay.io/maistra-dev
 
-$(KIND):
-	@GOBIN=$(LOCALBIN) go install -mod=readonly sigs.k8s.io/kind@v0.26.0
+.PHONY: docker-build
+docker-build: build ## Builds container image
+	$(CONTAINER_CLI) build -t $(HUB)/federation-controller:$(TAG) -f $(PROJECT_DIR)/build/Dockerfile .
 
-$(CONTROLLER_GEN):
-	GOBIN=$(LOCALBIN) go install -mod=readonly sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+.PHONY: docker-push
+docker-push: ## Pushes container image to the registry
+	$(CONTAINER_CLI) push $(HUB)/federation-controller:$(TAG)
 
-.PHONY: clean
-clean:
-	@rm -rf $(LOCALBIN) $(PROJECT_DIR)/$(OUT_DIR)
+.PHONY: docker
+docker: docker-build docker-push ## Combines build and push targets
 
-##@ Code Gen
+## Code Gen
```

Was removing `@` intentional?

```suggestion
##@ Code Gen
```

---

Yes, these are implicit targets which we always invoke. On top of that you won't know how to call them as they're all behind variables.

---

Ah yes! That makes sense. Thanks.